lib/providers/hlr-lookups-provider.js

```diff
@@ -60,15 +77,258 @@ HLRLookupsProvider.prototype.hlrLookup = function hlrLookup(number) {
           logger.warn(logFunc('Received response not deemed successful for number' + number, response));
           return reject(response);
         }
-
       };
-
       hlrClient.submitSyncLookupRequest(responseHandler, number);
     }
 
   })
   .then(this.hlrExtractor.extract);
 };
 
-module.exports = HLRLookupsProvider;
+/**
+ * An async HLR lookup initiator the first sets the callbackUlr and then submits numbers for processing retuning a
+ * result of numbers that can be processed.
+ * @param numbers {string[]}  e164 formatted numbers to be processed
+ * @param callbackUrl {string} callbackURl to be used
+ * @param options {!Object} options object that must contain a callbackUrl
+ * @returns {Promise} reject if failure, otherwise a resolve with object result = {accepted: [], rejected:[], fromCache:[]}
+ * result.accepted is the total list of numbers accepted for hlr lookup
+ * results.rejected is the list of numbers rejected by lookup service
+ * results.fromCache is the list of numbers whose lookup results were retrieved from cache and were excluded from lookup call.
+ * Note that fromCache can contain entries from result.accepted. So the actual phone numbers sent to hlr lookups service are
+ * unique numbers from accepted and fromCache arrays.
+ */
+HLRLookupsProvider.prototype.asyncHlrLookup = function asyncHlrLookup(numbers, callbackUrl, options) {
+  var logger = this.logger;
+  var name = this.name;
+  var hlrClient = isProviderLookupEnabled ? this.hlrClient : mockHlrClient;
+  var logFunc = loggingFunc('asyncHlrLookup');
+  var validator = this.validator;
+  var uniqueId = uuid.v4();
+
+  // validator that does not throw error
+  var isNumberValid = function isNumberValid(number) {
```

Comment:
OK so, the problem to do that here is that validator.isValid is used elsewhere where I need the error to return it to the user. For instance: phoneNumberUtils.isValid accepts an array of numbers and returns entries with error so user can know why some of the numbers have failed the check (this has been published).